lam4-backend/gf-grammar/Lam4Eng.gf

```diff
@@ -0,0 +1,216 @@
+concrete Lam4Eng of Lam4 = open Prelude, Coordination in {
+  lincat
+    -- TypeDecl
+    -- RowTypeDecl
+    [RowTypeDecl] = {s : Str ; isEmpty : IsEmpty} ;
+    Metadata = LinMetadata ;
+    ListExpr = ListX0 ;
+    ListName = ListX0 ;
+  param
+    IsEmpty = Empty | NonEmpty ;
+    MyListSize = Zero | One | Many ;
+
+  oper
+    ListX0 : Type = ListX ** {size : MyListSize} ;
+
+    baseListX0 : ListX0 = {
+      s1, s2 = "" ;
+      size = Zero
+    } ;
+
+    consListX0 : SS -> ListX0 -> ListX0 = \x,xs -> case xs.size of {
+        Many => xs ** {
+          s1 = x.s ++ bindComma ++ xs.s1
+          } ;
+        One => xs ** {
+          s1 = x.s ;
+          size = Many
+          } ;
+        Zero => xs ** {
+          s2 = x.s ;
+          size = One
+          }
+      } ;
+
+    LinMetadata : Type = {s : Str ; isEmpty : IsEmpty} ;
+    linRowMd : LinMetadata -> Str = \md ->
+      case md.isEmpty of {
+        Empty => md.s ;
+        NonEmpty => "; i.e." ++ md.s
+      } ;
+    linTypeDeclMd : LinMetadata -> Str = \md ->
+      case md.isEmpty of {
+        Empty => "is a class" ++ md.s ;
+        NonEmpty => "is" ++ md.s
+      } ;
+    linType : SS -> Str = \s -> "of type" ++ s.s ;
+
+  lin
+    -- Placeholder, or skip some constructs?
+    EmptyS = {s = ""} ;
+    TypeDeclS td = td ;
+    ExprS expr = expr ;
+    EvalS expr = {s = "evaluate" ++ expr.s} ;
+    EvalWhetherS expr = {s = "evaluate whether" ++ expr.s} ;
+    AssignS name expr = {s = quote name.s ++ "is assigned the value" ++ quote expr.s} ;
+
+    -- Metadata
+    MkMetadata str = str ** {isEmpty = NonEmpty} ;
+    NoMetadata = {s = [] ; isEmpty = Empty} ;
+
+    -- Type declarations
+    MkRowTypeDecl md field typ = {s = bullet ++ "its" ++ field.s ++ linType typ ++ linRowMd md} ;
+    MkRowDecl md field = {s = bullet ++ "its" ++ field.s ++ linRowMd md} ;
+
+    -- These funs are automatically generated from cat [RowTypeDecl]{0} ;
+    -- : [RowTypeDecl]
+    BaseRowTypeDecl = {s = [] ; isEmpty = Empty} ;
+    -- : RowTypeDecl -> [RowTypeDecl] -> [RowTypeDecl]
+    ConsRowTypeDecl t ts =
+      let sep : Str = case ts.isEmpty of {
+                        Empty => [] ;
+                        NonEmpty => linebreak } ;
+       in {s = tab ++ t.s ++ sep ++ ts.s ; isEmpty = NonEmpty} ;
+
+    MkTypeDecl md name rtds = {
+      s = artIndef ++ name.s ++ linTypeDeclMd md ++ "." ++
+          case rtds.isEmpty of {
+            Empty => [] ;
+            NonEmpty => "Each" ++ name.s ++
+                     "has associated with it information like" ++
+                     linebreak ++ rtds.s
+          } ;
+      } ;
+
+  oper
+    bullet = "*" ;
```

Comment:
I wonder if it might make sense in the future to have something like a gf config file for stuff like what the tab and linebreak characters are